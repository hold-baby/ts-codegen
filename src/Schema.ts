import { getUseExtends, isArray, toCapitalCase } from "src/utils";
import { forEach, isEmpty, map, reduce } from "lodash";
import { IReference, ISchema } from "src/__types__/OpenAPI";
import { CustomType, Type } from "src/Type";
import { CustomSchema } from "src/__types__/types";
import { createRegister } from "src/Register";

export class Schema {
  private type: Type;

  constructor(register: ReturnType<typeof createRegister>) {
    this.type = new Type(register);
  }

  public convert(schema: CustomSchema, id?: string): CustomType {
    const name = id ? toCapitalCase(id) : id;
    const oneOf = (schema as ISchema).oneOf || (schema as ISchema).anyOf;
    if (oneOf) {
      return this.type.oneOf(map(oneOf, (v) => this.convert(v)));
    }

    const allOf = (schema as ISchema).allOf;
    if (allOf) {
      const { props, refs, useExtends } = this.handleAllOf(allOf, name);
      return this.type.object(props, refs, useExtends);
    }

    if (schema.items) {
      return this.type.array(this.handleItems(schema.items, name));
    }

    if (schema.$ref) {
      return this.type.ref((schema as IReference).$ref);
    }

    if (schema.enum) {
      return this.type.enum(schema.enum, name);
    }

    if (schema.type === "object" || schema.properties) {
      return schema.properties ? this.type.object(this.handleObject(schema, name)) : this.type.object("object");
    }

    if (schema.type === "string") {
      return this.type.string();
    }

    if (schema.type === "boolean") {
      return this.type.boolean();
    }

    if (schema.type === "integer" || schema.type === "number") {
      return this.type.number();
    }

    if (schema.type === "file") {
      return this.type.file();
    }

    return this.type.null();
  }

  private handleAllOf(schemas: Array<CustomSchema>, name?: string) {
    const refs: any[] = [];
    let props: any = {};
    let useExtends = getUseExtends(schemas);

    forEach(schemas, (schema) => {
      if (schema.$ref) {
        refs.push(this.convert(schema, name));
      } else if (!isEmpty(schema)) {
        props = this.convert(schema, name);
      }
    });

    return {
      refs,
      props,
      useExtends,
    };
  }

  private handleItems(items: CustomSchema | IReference | CustomSchema[], name?: string): CustomType | CustomType[] {
    if (isArray(items)) {
      return map(items, (v) => this.handleItems(v, name)) as CustomType[];
    }
    return this.convert(items, name);
  }

  private handleObject(schema: CustomSchema, name?: string): { [key: string]: CustomType } {
    return reduce(
      schema.properties,
      (res, v, k) => {
        const isRequired = (v as CustomSchema)?.required || schema.required?.includes(k);
        return {
          ...res,
          [`${k}${isRequired ? "" : "?"}`]: this.convert(v, `${name}${toCapitalCase(k)}`),
        };
      },
      {},
    );
  }
}
