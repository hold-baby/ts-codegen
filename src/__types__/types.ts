import { CustomType } from "src/Type";
import { Operation, Parameter, Path, Schema, Spec } from "swagger-schema-official";
import { IComponents, IOpenAPI, IOperation, IPathItem, ISchema, TParameter } from "src/__types__/OpenAPI";

export type RequestType = { [key: string]: CustomType };

export interface IClientConfig {
  url: string;
  method: string;
  TResp: CustomType | undefined;
  TReq: RequestType | { [key: string]: RequestType } | undefined;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  bodyParams?: string[]; // for backward capability, only used in swagger version 2.0
  contentType?: string;
  deprecated?: boolean;
}

export type CustomSchema = Schema | ISchema;
export type CustomReference = {
  $ref: string;
};
export type CustomParameters = Operation["parameters"] | IOperation["parameters"];
export type CustomResponses = Spec["responses"] | IComponents["responses"];
export type CustomParameter = Parameter | TParameter;
export type CustomPath = Path | IPathItem;
export type CustomPaths = Spec["paths"] | IOpenAPI["paths"];
export type CustomOperation = Operation | IOperation;
