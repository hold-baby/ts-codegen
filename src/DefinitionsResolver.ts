import { Schema } from "swagger-schema-official";
import { addPrefixForInterface, arrayToObject, isNumber, toCapitalCase, toTypes } from "./utils";
import { compact, Dictionary, forEach, includes, replace, some } from "lodash";
import { SchemaResolver2 } from "src/swagger/v3/SchemaResolver2";

// TODO: 1. Handle required params.
// TODO: handle `in: fromData`
// TODO: handle `in schema`
// TODO: 确认不同 endpoint 是否都会生成 swagger

const ENUM_SUFFIX = `#EnumTypeSuffix`;

export function generateEnums(data: Dictionary<any>, key: string) {
  if (!data) {
    return "";
  }

  const enums = data[key];
  const hasNumber = some(enums, (v) => isNumber(v));
  const enumName = replace(key, ENUM_SUFFIX, "");
  return hasNumber
    ? `export type ${enumName} = ${enums.map((item: string | number) => JSON.stringify(item)).join("|")}`
    : `export enum ${enumName} ${JSON.stringify(arrayToObject(enums)).replace(/:/gi, "=")}`;
}

export class DefinitionsResolver {
  public resolvedDefinitions: Dictionary<any> = {};

  static of(definitions?: { [definitionsName: string]: Schema }) {
    return new DefinitionsResolver(definitions);
  }

  constructor(private definitions?: { [definitionsName: string]: Schema }) {}

  scan = () => {
    const r = SchemaResolver2.of((k, v) => {
      this.resolvedDefinitions[k] = v;
    });

    forEach(this.definitions, (v, k) =>
      r.resolve({
        ...v,
        _name: k,
        _propKey: k,
      }),
    );

    return this;
  };

  toDeclarations = (): string[] => {
    const resolvedDefinitions = this.resolvedDefinitions;
    const arr = Object.keys(resolvedDefinitions)
      .sort()
      .map((key) => {
        if (includes(key, ENUM_SUFFIX)) {
          return generateEnums(resolvedDefinitions, key);
        }

        if (resolvedDefinitions[key] === "object") {
          return `export interface ${addPrefixForInterface(toCapitalCase(key))} {[key:string]:any}`;
        }
        const val = toTypes(resolvedDefinitions[key]);
        if (val) {
          return `export interface ${addPrefixForInterface(toCapitalCase(key))} ${val}`;
        }
      });

    return compact(arr);
  };
}
