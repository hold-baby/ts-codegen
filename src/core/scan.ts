import { IReference } from "src/v3/OpenAPI";
import { CustomSchema, Ref, Register } from "src/core/Type";
import { keys } from "lodash";
import { getUseExtends, Schema } from "src/core/Schema";
import { prettifyCode, toCapitalCase } from "src/core/utils";

const getDeclarationType = (schema: CustomSchema) => {
  if (schema.type === "object" || schema.properties || (schema.allOf && getUseExtends(schema.allOf))) {
    return "interface";
  }
  return "type";
};

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const scan = (schemas: { [k: string]: CustomSchema | IReference }) => {
  const schemaHandler = new Schema();

  keys(schemas).forEach((k) => {
    const name = toCapitalCase(k);
    const type = schemaHandler.convert(schemas[k], name);
    Register.setType(name, type);
    Register.setPrefix(name, getDeclarationType(schemas[k]));
  });

  for (let name in Register.refs) {
    (Register.refs[name] as Ref).rename(addPrefix(name));
  }

  let output = "";
  keys(Register.decls)
    .sort()
    .forEach((k) => {
      output =
        output +
        `export ${Register.prefixes[k]} ${addPrefix(k)} ${
          Register.prefixes[k] === "interface" ? "" : "="
        } ${Register.decls[k].toType()}${Register.prefixes[k] === "type" ? ";" : ""}\n\n`;
    });

  return prettifyCode(output);
};
