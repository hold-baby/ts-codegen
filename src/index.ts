export * from "src/v2/ClientConfigs";
export * from "src/v3/ClientConfigsV3";
export * from "src/core/utils";
