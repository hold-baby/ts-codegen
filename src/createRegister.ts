// 利用闭包持有状态（私有变量）
import { CustomType, Ref } from "src/Type";
import { Parameter } from "swagger-schema-official";
import { IReference, IRequestBody, IResponse } from "src/__types__/OpenAPI";
import { get, set } from "lodash";

export enum DeclKinds {
  interface = "interface",
  type = "type",
  enum = "enum",
}

const withPrefix = (name: string, kind: string) => {
  switch (kind) {
    case "interface":
      return `I${name}`;
    case "type":
      return `T${name}`;
    default:
      return name;
  }
};

export interface IStore {
  decls: {
    [id: string]: {
      type: CustomType;
      kind: string;
      name: string;
    };
  };
  refs: { [id: string]: Ref };
  parameters: {
    [id: string]: Parameter;
  };
  responses: {
    [id: string]: Response | IResponse | IReference;
  };
  requestBodies: {
    [id: string]: IReference | IRequestBody;
  };
}

export const createRegister = () => {
  const store: IStore = {
    decls: {},
    refs: {},
    parameters: {},
    responses: {},
    requestBodies: {},
  };

  return {
    getDecls() {
      return store.decls;
    },

    setDecl: (id: string, type: CustomType, kind: string) => {
      store.decls[id] = {
        type,
        kind,
        name: withPrefix(id, kind),
      };
    },

    setRef: (id: string) => {
      if (store.refs[id]) {
        return store.refs[id];
      }

      const type = new Ref(id);
      store.refs[id] = type;

      return type;
    },

    getData(paths: string[]) {
      return get(store, paths);
    },

    setData(paths: string[], data: any) {
      return set(store, paths, data);
    },

    renameAllRefs: (cb: (newName: string) => string) => {
      for (let name in store.refs) {
        store.refs[name].rename(cb(name));
      }
    },
  };
};
