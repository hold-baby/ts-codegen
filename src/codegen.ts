import { getFilename, prettifyCode, testJSON } from "src/utils";
import axios from "axios";
import { isEmpty, map } from "lodash";
import * as fs from "fs";
import * as path from "path";
import { IOpenAPI } from "src/__types__/OpenAPI";
import { Spec } from "swagger-schema-official";
import { getInputs, scan } from "src/scan";
import { ERROR_MESSAGES } from "src/constants";

export const getCodegenConfig = () => {
  const codegenConfigPath = path.resolve("ts-codegen.config.json");
  return fs.existsSync(codegenConfigPath)
    ? require(codegenConfigPath)
    : {
        output: ".output",
        actionCreatorImport: "",
        clients: [],
        options: {
          typeWithPrefix: false,
          backwardCompatible: false,
        },
      };
};

export const codegen = () => {
  const { output, actionCreatorImport, timeout, data, clients, options } = getCodegenConfig();

  const writeSpecToFile = (spec: IOpenAPI | Spec) => {
    if (!spec) {
      return;
    }
    const fileStr = `${actionCreatorImport} ${scan(spec, options)}`;
    const { basePath } = getInputs(spec);
    write(output, `./${getFilename(basePath)}`, fileStr);
  };

  if (!isEmpty(data)) {
    data.map((file: string) => {
      const specStr = fs.readFileSync(file, "utf8");
      const spec = testJSON(specStr, ERROR_MESSAGES.INVALID_JSON_FILE_ERROR);

      writeSpecToFile(spec);
    });
  }

  if (!isEmpty(clients)) {
    fetchSwaggerJSON(clients, timeout).then((results: any[]) => {
      results.forEach((spec: IOpenAPI | Spec) => {
        writeSpecToFile(spec);
      });
    });
  }
};

const write = (output: string, filename: string, str: string) => {
  if (!fs.existsSync(output)) {
    fs.mkdirSync(output);
  }

  fs.writeFileSync(path.resolve(output, `./${filename}.ts`), prettifyCode(str), "utf-8");
};

const fetchSwaggerJSON = (clients: string[] = [], timeout: number = 10 * 1000) => {
  const instance = axios.create({
    timeout,
  });

  return Promise.all(
    map(clients, (client) => {
      instance
        .get(client)
        .then((response) => response.data)
        .catch((error) => {
          console.error(`${error.code}: ${ERROR_MESSAGES.FETCH_CLIENT_FAILED_ERROR}`);
        });
    }),
  );
};
