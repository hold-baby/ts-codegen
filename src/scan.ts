import { IOpenAPI, IServer } from "src/__types__/OpenAPI";
import { CustomType } from "src/Type";
import { compact, get, isEmpty, keys, mapValues, sortBy } from "lodash";
import { Schema } from "src/Schema";
import { getUseExtends, prettifyCode, setDeprecated, toCapitalCase, toTypes } from "src/utils";
import { Spec } from "swagger-schema-official";
import { CustomReference, CustomSchema, IClientConfig, RequestType } from "src/__types__/types";
import { createRegister, DeclKinds, IStore } from "src/createRegister";
import { parse } from "url";
import { getClientConfigsV2, getClientConfigsV3 } from "src/createClientConfigs";

enum DataType {
  openapi,
  swagger,
}

interface ScanOptions {
  typeWithPrefix?: boolean; // Will keep prefix('I' for interface, 'T' for type) in types when it sets true
  backwardCompatible?: boolean; // Not recommend, only if you want backward capability. This option will help to keep operationId and method name as before when it sets true. This option is only worked with swagger version 2.0.
}

export const scan = (data: Spec | IOpenAPI, options?: ScanOptions) => {
  const register = createRegister(options?.typeWithPrefix);
  const schemaHandler = new Schema(register);
  const { dataType, basePath, paths, schemas, parameters, responses, requestBodies } = getInputs(data);

  keys(schemas).forEach((k) => {
    const name = toCapitalCase(k);
    register.setDecl(name, schemaHandler.convert(schemas[k], name), getDeclarationType(schemas[k]));
  });

  register.setData(["parameters"], parameters);
  register.setData(["responses"], responses);
  register.setData(["requestBodies"], requestBodies);

  let clientConfigs: IClientConfig[] =
    dataType === DataType.swagger
      ? getClientConfigsV2(paths, basePath, register, options?.backwardCompatible)
      : getClientConfigsV3(paths, basePath, register);

  const decls = register.getDecls();
  if (options?.typeWithPrefix) {
    register.renameAllRefs((key) => decls[key].name);
  }

  return print(clientConfigs, decls);
};

const isOpenApi = (v: any): v is IOpenAPI => v.openapi;

export const getInputs = (data: Spec | IOpenAPI) => {
  if (isOpenApi(data)) {
    return {
      dataType: DataType.openapi,
      basePath: getBasePathFromServers(data?.servers),
      paths: data.paths,
      schemas: data.components?.schemas as { [key: string]: CustomSchema | CustomReference },
      parameters: data.components?.parameters,
      responses: data.components?.responses,
      requestBodies: data.components?.requestBodies,
    };
  }
  return {
    dataType: DataType.swagger,
    basePath: data.basePath || "",
    paths: data.paths,
    schemas: data.definitions as { [key: string]: CustomSchema },
    parameters: data.parameters,
    responses: data.responses,
    requestBodies: null,
  };
};

const print = (clientConfigs: IClientConfig[], decls: IStore["decls"]) => {
  return prettifyCode(`${printRequest(clientConfigs)} \n\n ${printTypes(decls)}`);
};

function printRequest(clientConfigs: IClientConfig[]): string {
  const configs = sortBy(clientConfigs, (o) => o.operationId);

  return configs
    .map((v) => {
      const toUrl = () => `url: \`${v.url}\`,`;
      const toMethod = () => `method: "${v.method}",`;
      const toRequestBody = () => {
        if (!isEmpty(v.bodyParams)) {
          // TODO: refactor code
          return `data: ${v.bodyParams!.length > 1 ? `{${v.bodyParams!.join(",")}}` : v.bodyParams![0]},`;
        }
        return v.contentType ? "data: requestBody," : "";
      };
      const toQueryParams = () => {
        const params = toRequestParams(v.queryParams);
        return params ? `params: ${params},` : "";
      };
      const toHeaders = () => (v.contentType ? `headers: {"Content-Type": '${v.contentType}'},` : "");
      const toGenerators = () => {
        const types = compact([generateTReq(v.TReq), v.TResp?.toType(false)]).join(",");
        return types ? `<${types}>` : "";
      };
      const toRequestInputs = () => {
        const getRequestBody = () => {
          if (!isEmpty(v.bodyParams)) {
            return v.bodyParams!;
          }
          return v.contentType ? ["requestBody"] : "";
        };
        const list = compact([...v.pathParams, ...v.queryParams, ...getRequestBody()]);
        return isEmpty(list) ? "" : toRequestParams(list);
      };

      return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction${toGenerators()}("${
        v.operationId
      }", (${toRequestInputs()}) => ({${toUrl()}${toMethod()}${toRequestBody()}${toQueryParams()}${toHeaders()}})
);
`;
    })
    .join("\n\n");
}

const printTypes = (decls: IStore["decls"]): string => {
  return keys(decls)
    .sort()
    .map((k) => {
      const expr = decls[k].kind === DeclKinds.type ? "=" : "";
      const semi = decls[k].kind === DeclKinds.type ? ";" : "";
      return `export ${decls[k].kind} ${decls[k].name} ${expr} ${decls[k].type.toType()}${semi}`;
    })
    .join("\n\n");
};

function generateTReq(TReq: IClientConfig["TReq"]) {
  if (isEmpty(TReq)) {
    return "";
  }

  function gen(obj: IClientConfig["TReq"]): string {
    return toTypes(
      mapValues(obj, (v) => {
        if (!v.toType) {
          return gen(v as RequestType);
        }
        return (v as CustomType).toType(false);
      }),
    );
  }

  return gen(TReq);
}

const toRequestParams = (data: string[]) =>
  !isEmpty(data)
    ? `{
${data.join(",\n")}
}`
    : undefined;

const getDeclarationType = (schema: CustomSchema) => {
  if (schema?.type === "object" || schema?.properties || (schema?.allOf && getUseExtends(schema?.allOf))) {
    return DeclKinds.interface;
  }
  return DeclKinds.type;
};

const getBasePathFromServers = (servers?: IServer[]): string => {
  if (isEmpty(servers)) {
    return "";
  }
  const server = servers![0];
  if (server?.variables) {
    const basePath = get(server, "variables.basePath.default");
    return basePath ? `/${basePath}` : "";
  }
  return parse(server?.url)?.pathname || "";
};
