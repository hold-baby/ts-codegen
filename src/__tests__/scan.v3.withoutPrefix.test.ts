import swaggerV3 from "examples/openapi.json";
import { scan } from "src/scan";
import { IOpenAPI } from "src/__types__/OpenAPI";

describe("swagger v3", () => {
  it("should handle basic schemas without prefix in type name", () => {
    expect(scan(swaggerV3 as IOpenAPI)).toMatchSnapshot();
  });
});
