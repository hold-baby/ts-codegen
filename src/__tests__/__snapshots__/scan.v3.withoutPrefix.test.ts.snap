// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v3 should handle basic schemas without prefix in type name 1`] = `
"export const addPet = createRequestAction<
  {
    requestBody: NewPet;
  },
  Pet
>(\\"addPet\\", ({ requestBody }) => ({
  url: \`/api/v3/pets\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createItem = createRequestAction<
  {
    item?: Item;
  },
  Message
>(\\"createItem\\", ({ item }) => ({
  url: \`/api/v3/items\`,
  method: \\"post\\",
  params: {
    item,
  },
}));

export const deletePet = createRequestAction<{
  id: number;
}>(\\"deletePet\\", ({ id }) => ({ url: \`/api/v3/pets/\${id}\`, method: \\"delete\\" }));

export const findPetById = createRequestAction<
  {
    id: number;
  },
  Pet
>(\\"findPetById\\", ({ id }) => ({ url: \`/api/v3/pets/\${id}\`, method: \\"get\\" }));

export const findPets = createRequestAction<
  {
    limit?: number;
    status: keyof typeof FindPetsStatus[];
    tags?: keyof typeof FindPetsTags[];
  },
  Pet[]
>(\\"findPets\\", ({ tags, limit, status }) => ({
  url: \`/api/v3/pets\`,
  method: \\"get\\",
  params: {
    tags,
    limit,
    status,
  },
}));

export const formMultipartWithFile = createRequestAction<{
  requestBody: { data?: Pet; file: string; slice?: string[]; string?: string };
}>(\\"formMultipartWithFile\\", ({ requestBody }) => ({
  url: \`/api/v3/demo/forms/multipart\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const formMultipartWithFiles = createRequestAction<{
  requestBody: { files: string[] };
}>(\\"formMultipartWithFiles\\", ({ requestBody }) => ({
  url: \`/api/v3/demo/forms/multipart-with-files\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const formUrlEncoded = createRequestAction<{
  requestBody: RequestForm;
}>(\\"formUrlEncoded\\", ({ requestBody }) => ({
  url: \`/api/v3/demo/forms/url-encoded\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/x-www-form-urlencoded\\" },
}));

export const getItems = createRequestAction<
  {
    limit?: number;
    page?: number;
    sort?: string[];
  },
  Items
>(\\"getItems\\", ({ page, limit, sort }) => ({
  url: \`/api/v3/items\`,
  method: \\"get\\",
  params: {
    page,
    limit,
    sort,
  },
}));

export const postBody = createRequestAction<{
  requestBody: Pet;
}>(\\"postBody\\", ({ requestBody }) => ({
  url: \`/api/v3/body\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePet = createRequestAction<{
  requestBody?: Cat | Dog;
}>(\\"updatePet\\", ({ requestBody }) => ({
  url: \`/api/v3/pets\`,
  method: \\"patch\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export interface Cat extends Pet {
  age?: number;
  hunts?: boolean;
}

export interface Category {
  id?: number;
  name?: string;
}

export interface Dog extends Pet {
  bark?: boolean;
  breed?: keyof typeof DogBreed;
}

export enum DogBreed {
  \\"Dingo\\" = \\"Dingo\\",
  \\"Husky\\" = \\"Husky\\",
  \\"Retriever\\" = \\"Retriever\\",
  \\"Shepherd\\" = \\"Shepherd\\",
}

export interface Error {
  code: number;
  message: string;
}

export enum FindPetsStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export enum FindPetsTags {
  \\"z\\" = \\"z\\",
  \\"b\\" = \\"b\\",
  \\"c\\" = \\"c\\",
  \\"a\\" = \\"a\\",
}

export interface Item {
  color?: string;
  id?: string;
  name?: string;
  price?: string;
}

export type Items = Item[];

export interface Message {
  message?: string;
}

export interface NewPet {
  name: string;
  tag?: string;
}

export interface Pet extends NewPet {
  categories?: Category[];
  id: number;
}

export type Pets = Pet[];

export interface RequestForm {
  data: Pet;
  slice: string[];
  string: string;
}
"
`;
